File: app/src/main/java/com/softbankrobotics/qisdktutorials/ui/categories/CategoriesRobot.java
package com.softbankrobotics.qisdktutorials.ui.categories;

import android.support.annotation.NonNull;
import android.util.Log;

import com.aldebaran.qi.Consumer;
import com.aldebaran.qi.Future;
import com.aldebaran.qi.sdk.QiContext;
import com.aldebaran.qi.sdk.QiSDK;
import com.aldebaran.qi.sdk.RobotLifecycleCallbacks;
import com.aldebaran.qi.sdk.builder.DiscussBuilder;
import com.aldebaran.qi.sdk.builder.SayBuilder;
import com.aldebaran.qi.sdk.builder.TopicBuilder;
import com.aldebaran.qi.sdk.object.conversation.Bookmark;
import com.aldebaran.qi.sdk.object.conversation.Discuss;
import com.aldebaran.qi.sdk.object.conversation.QiChatVariable;
import com.aldebaran.qi.sdk.object.conversation.Topic;
import com.aldebaran.qi.sdk.object.conversation.TopicStatus;
import com.softbankrobotics.qisdktutorials.R;
import com.softbankrobotics.qisdktutorials.model.data.Tutorial;
import com.softbankrobotics.qisdktutorials.model.data.TutorialCategory;
import com.softbankrobotics.qisdktutorials.model.data.TutorialLevel;

/**
 * The robot for the tutorial categories.
 */
class CategoriesRobot implements CategoriesContract.Robot, RobotLifecycleCallbacks {

    private static final String TAG = "CategoriesRobot";

    private static final String LEVEL_BASIC = "basic";
    private static final String LEVEL_ADVANCED = "advanced";

    private final CategoriesContract.Presenter presenter;
    private TopicStatus talkTopicStatus;
    private TopicStatus moveTopicStatus;
    private TopicStatus smartTopicStatus;
    private Discuss discuss;
    private Future<String> discussFuture;
    private TutorialCategory selectedCategory = TutorialCategory.TALK;
    private TutorialLevel selectedLevel = TutorialLevel.BASICS;
    private QiChatVariable levelVariable;

    CategoriesRobot(CategoriesContract.Presenter presenter) {
        this.presenter = presenter;
    }

    @Override
    public void register(CategoriesActivity activity) {
        QiSDK.register(activity, this);
    }

    @Override
    public void unregister(CategoriesActivity activity) {
        QiSDK.unregister(activity, this);
    }

    @Override
    public void stopDiscussion(final Tutorial tutorial) {
        if (discussFuture != null) {
            discussFuture.thenConsume(new Consumer<Future<String>>() {
                @Override
                public void consume(Future<String> future) throws Throwable {
                    if (future.isCancelled()) {
                        presenter.goToTutorial(tutorial);
                    }
                }
            });
            discussFuture.requestCancellation();
        } else {
            presenter.goToTutorial(tutorial);
        }
    }

    @Override
    public void selectTopic(final TutorialCategory category) {
        selectedCategory = category;

        boolean topicsAreReady = talkTopicStatus != null && moveTopicStatus != null && smartTopicStatus != null;
        if (topicsAreReady) {
            enableTopic(category);
        }
    }

    @Override
    public void selectLevel(TutorialLevel level) {
        selectedLevel = level;

        if (levelVariable != null) {
            enableLevel(level);
        }
    }

    @Override
    public void onRobotFocusGained(QiContext qiContext) {
        SayBuilder.with(qiContext)
                .withText(qiContext.getString(R.string.categories_intro_sentence))
                .build()
                .run();

        final Topic commonTopic = TopicBuilder.with(qiContext)
                .withResource(R.raw.common)
                .build();

        Topic talkTopic = TopicBuilder.with(qiContext)
                .withResource(R.raw.talk_tutorials)
                .build();

        Topic moveTopic = TopicBuilder.with(qiContext)
                .withResource(R.raw.move_tutorials)
                .build();

        Topic smartTopic = TopicBuilder.with(qiContext)
                .withResource(R.raw.smart_tutorials)
                .build();

        discuss = DiscussBuilder.with(qiContext)
                .withTopic(commonTopic, talkTopic, moveTopic, smartTopic)
                .build();

        talkTopicStatus = discuss.topicStatus(talkTopic);
        moveTopicStatus = discuss.topicStatus(moveTopic);
        smartTopicStatus = discuss.topicStatus(smartTopic);

        levelVariable = discuss.variable("level");

        enableLevel(selectedLevel);
        enableTopic(selectedCategory);

        discuss.setOnBookmarkReachedListener(new Discuss.OnBookmarkReachedListener() {
            @Override
            public void onBookmarkReached(Bookmark bookmark) {
                String bookmarkName = bookmark.getName();
                switch (bookmarkName) {
                    case "talk":
                        presenter.loadTutorials(TutorialCategory.TALK);
                        selectTopic(TutorialCategory.TALK);
                        break;
                    case "move":
                        presenter.loadTutorials(TutorialCategory.MOVE);
                        selectTopic(TutorialCategory.MOVE);
                        break;
                    case "smart":
                        presenter.loadTutorials(TutorialCategory.SMART);
                        selectTopic(TutorialCategory.SMART);
                        break;
                    case "basics":
                        presenter.loadTutorials(TutorialLevel.BASICS);
                        selectLevel(TutorialLevel.BASICS);
                        break;
                    case "advanced":
                        presenter.loadTutorials(TutorialLevel.ADVANCED);
                        selectLevel(TutorialLevel.ADVANCED);
                        break;
                }
            }
        });

        discussFuture = discuss.async().run();

        discussFuture.andThenConsume(new Consumer<String>() {
            @Override
            public void consume(String tutorialDiscussId) throws Throwable {
                presenter.goToTutorialForDiscussId(tutorialDiscussId);
            }
        });
    }

    @Override
    public void onRobotFocusLost() {
        if (discuss != null) {
            discuss.setOnBookmarkReachedListener(null);
            discuss = null;
        }
        discussFuture = null;
        talkTopicStatus = null;
        moveTopicStatus = null;
        smartTopicStatus = null;
    }

    @Override
    public void onRobotFocusRefused(String reason) {
        Log.i(TAG, "onRobotFocusRefused: " + reason);
    }

    /**
     * Enable the topic corresponding to the specified tutorial category.
     * @param category the tutorial category
     */
    private void enableTopic(final TutorialCategory category) {
        Future<Void> talkFuture = talkTopicStatus.async().setEnabled(false);
        Future<Void> moveFuture = moveTopicStatus.async().setEnabled(false);
        Future<Void> smartFuture = smartTopicStatus.async().setEnabled(false);

        Future.waitAll(talkFuture, moveFuture, smartFuture)
                .andThenConsume(new Consumer<Void>() {
                    @Override
                    public void consume(Void ignore) throws Throwable {
                        switch (category) {
                            case TALK:
                                talkTopicStatus.setEnabled(true);
                                break;
                            case MOVE:
                                moveTopicStatus.setEnabled(true);
                                break;
                            case SMART:
                                smartTopicStatus.setEnabled(true);
                                break;
                            default:
                                throw new IllegalArgumentException("Unknown tutorial category: " + category);
                        }
                    }
                });
    }

    /**
     * Enable the specified level.
     * @param level the tutorial level
     */
    private void enableLevel(TutorialLevel level) {
        String value = levelValueFromLevel(level);
        levelVariable.async().setValue(value);
    }

    /**
     * Provides the level variable value from the specified tutorial level.
     * @param level the tutorial level
     * @return The level variable value.
     */
    @NonNull
    private String levelValueFromLevel(TutorialLevel level) {
        String value;
        switch (level) {
            case BASICS:
                value = LEVEL_BASIC;
                break;
            case ADVANCED:
                value = LEVEL_ADVANCED;
                break;
            default:
                throw new IllegalArgumentException("Unknown tutorial level: " + level);
        }
        return value;
    }
}

File: app/src/main/java/com/softbankrobotics/qisdktutorials/ui/categories/CategoriesRobot.kt
/*
 * Copyright (C) 2018 Softbank Robotics Europe
 * See COPYING for the license
 */

package com.softbankrobotics.qisdktutorials.ui.categories

import android.support.annotation.StringRes
import android.util.Log

import com.aldebaran.qi.Future
import com.aldebaran.qi.sdk.QiContext
import com.aldebaran.qi.sdk.QiSDK
import com.aldebaran.qi.sdk.RobotLifecycleCallbacks
import com.aldebaran.qi.sdk.builder.ChatBuilder
import com.aldebaran.qi.sdk.builder.QiChatbotBuilder
import com.aldebaran.qi.sdk.builder.SayBuilder
import com.aldebaran.qi.sdk.builder.TopicBuilder
import com.aldebaran.qi.sdk.`object`.conversation.QiChatVariable
import com.aldebaran.qi.sdk.`object`.conversation.QiChatbot
import com.aldebaran.qi.sdk.`object`.conversation.TopicStatus
import com.softbankrobotics.qisdktutorials.R
import com.softbankrobotics.qisdktutorials.model.data.Tutorial
import com.softbankrobotics.qisdktutorials.model.data.TutorialCategory
import com.softbankrobotics.qisdktutorials.model.data.TutorialLevel

private const val TAG = "CategoriesRobot"

private const val LEVEL_BASIC = "basic"
private const val LEVEL_ADVANCED = "advanced"

/**
 * The robot for the tutorial categories.
 */
internal class CategoriesRobot(private val presenter: CategoriesContract.Presenter) : CategoriesContract.Robot, RobotLifecycleCallbacks {
    private var talkTopicStatus: TopicStatus? = null
    private var moveTopicStatus: TopicStatus? = null
    private var smartTopicStatus: TopicStatus? = null
    private var qiChatbot: QiChatbot? = null
    private var chatFuture: Future<Void>? = null
    private var selectedCategory = TutorialCategory.TALK
    private var selectedLevel = TutorialLevel.BASIC
    private var levelVariable: QiChatVariable? = null
    private var isFirstIntro = true

    override fun register(activity: CategoriesActivity) {
        QiSDK.register(activity, this)
    }

    override fun unregister(activity: CategoriesActivity) {
        QiSDK.unregister(activity, this)
    }

    override fun stopDiscussion(tutorial: Tutorial) {
        val chatFuture = chatFuture
        if (chatFuture != null) {
            chatFuture.thenConsume {
                if (it.isCancelled) {
                    presenter.goToTutorial(tutorial)
                }
            }
            chatFuture.requestCancellation()
        } else {
            presenter.goToTutorial(tutorial)
        }
        this.chatFuture = chatFuture
    }

    override fun selectTopic(category: TutorialCategory) {
        selectedCategory = category

        val topicsAreReady = talkTopicStatus != null && moveTopicStatus != null && smartTopicStatus != null
        if (topicsAreReady) {
            enableTopic(category)
        }
    }

    override fun selectLevel(level: TutorialLevel) {
        selectedLevel = level

        if (levelVariable != null) {
            enableLevel(level)
        }
    }

    override fun onRobotFocusGained(qiContext: QiContext) {
        SayBuilder.with(qiContext)
                .withText(qiContext.getString(introSentenceRes()))
                .build()
                .run()

        isFirstIntro = false

        val commonTopic = TopicBuilder.with(qiContext)
                .withResource(R.raw.common)
                .build()

        val talkTopic = TopicBuilder.with(qiContext)
                .withResource(R.raw.talk_tutorials)
                .build()

        val moveTopic = TopicBuilder.with(qiContext)
                .withResource(R.raw.move_tutorials)
                .build()

        val smartTopic = TopicBuilder.with(qiContext)
                .withResource(R.raw.smart_tutorials)
                .build()

        val qiChatbot = QiChatbotBuilder.with(qiContext)
                .withTopics(listOf(commonTopic, talkTopic, moveTopic, smartTopic))
                .build()
                .also { this.qiChatbot = it }

        val chat = ChatBuilder.with(qiContext)
                .withChatbot(qiChatbot)
                .build()

        talkTopicStatus = qiChatbot.topicStatus(talkTopic)
        moveTopicStatus = qiChatbot.topicStatus(moveTopic)
        smartTopicStatus = qiChatbot.topicStatus(smartTopic)

        levelVariable = qiChatbot.variable("level")

        enableLevel(selectedLevel)
        enableTopic(selectedCategory)

        qiChatbot.addOnBookmarkReachedListener { bookmark ->
            when (bookmark.name) {
                "talk" -> {
                    presenter.loadTutorials(TutorialCategory.TALK)
                    selectTopic(TutorialCategory.TALK)
                }
                "move" -> {
                    presenter.loadTutorials(TutorialCategory.MOVE)
                    selectTopic(TutorialCategory.MOVE)
                }
                "smart" -> {
                    presenter.loadTutorials(TutorialCategory.SMART)
                    selectTopic(TutorialCategory.SMART)
                }
                "basic" -> {
                    presenter.loadTutorials(TutorialLevel.BASIC)
                    selectLevel(TutorialLevel.BASIC)
                }
                "advanced" -> {
                    presenter.loadTutorials(TutorialLevel.ADVANCED)
                    selectLevel(TutorialLevel.ADVANCED)
                }
            }
        }

        qiChatbot.addOnEndedListener { presenter.goToTutorialForQiChatbotId(it) }
        this.qiChatbot = qiChatbot
        chatFuture = chat.async().run()
    }

    override fun onRobotFocusLost() {
        qiChatbot?.let {
            it.removeAllOnBookmarkReachedListeners()
            it.removeAllOnEndedListeners()
            qiChatbot = null
        }
        chatFuture = null
        talkTopicStatus = null
        moveTopicStatus = null
        smartTopicStatus = null
    }

    override fun onRobotFocusRefused(reason: String) {
        Log.i(TAG, "onRobotFocusRefused: $reason")
    }

    /**
     * Enable the topic corresponding to the specified tutorial category.
     * @param category the tutorial category
     */
    private fun enableTopic(category: TutorialCategory) {
        val talkFuture = talkTopicStatus?.async()?.setEnabled(false)
        val moveFuture = moveTopicStatus?.async()?.setEnabled(false)
        val smartFuture = smartTopicStatus?.async()?.setEnabled(false)

        Future.waitAll(talkFuture, moveFuture, smartFuture)
                .andThenConsume {
                    when (category) {
                        TutorialCategory.TALK -> talkTopicStatus?.enabled = true
                        TutorialCategory.MOVE -> moveTopicStatus?.enabled = true
                        TutorialCategory.SMART -> smartTopicStatus?.enabled = true
                    }
                }
    }

    /**
     * Enable the specified level.
     * @param level the tutorial level
     */
    private fun enableLevel(level: TutorialLevel) {
        val value = levelValueFromLevel(level)
        levelVariable?.async()?.setValue(value)
    }

    /**
     * Provides the level variable value from the specified tutorial level.
     * @param level the tutorial level
     * @return The level variable value.
     */
    private fun levelValueFromLevel(level: TutorialLevel): String {
        return when (level) {
            TutorialLevel.BASIC -> LEVEL_BASIC
            TutorialLevel.ADVANCED -> LEVEL_ADVANCED
        }
    }

    @StringRes
    private fun introSentenceRes(): Int {
        return if (isFirstIntro) R.string.categories_intro_sentence else R.string.categories_intro_sentence_variant
    }

}
